h/services/user_delete.py

```diff
@@ -1,70 +1,348 @@
-                    sa.select(
-                        sa.func.count(Annotation.id)  # pylint:disable=not-callable
+                num_annotations=self.db.scalar(
+                    select(
+                        func.count(Annotation.id)  # pylint:disable=not-callable
                     ).where(Annotation.userid == user.userid)
                 ),
             )
         )
 
-        self._db.delete(user)
+    def purge_deleted_users(self, limit=1000):
+        """Incrementally purge data of users who've been marked as deleted."""
+        jobs = self.job_queue.get(Job.JobName.PURGE_USER, limit)
+
+        if not jobs:
+            return
+
+        completed_jobs = []
+        purger = UserPurger(self.db, self.job_queue, LimitedWorker(self.db, limit))
+
+        for job in jobs:
+            userid = job.kwargs.get("userid")
+
+            if not userid:
+                log.info("Invalid '%s' job: %s", Job.JobName.PURGE_USER, job)
+                completed_jobs.append(job)
+                continue
+
+            user = self.user_svc.fetch(userid)
+
+            if user is None:
+                log.info("Couldn't fetch user: %s", userid)
+                completed_jobs.append(job)
+                continue
+
+            log.info("Purging user: %s", user.userid)
+
+            try:
+                purger.delete_authtickets(user)
+                purger.delete_tokens(user)
+                purger.delete_flags(user)
+                purger.delete_featurecohort_memberships(user)
+                purger.delete_annotations(user)
+                purger.delete_groups(user)
+                purger.delete_group_memberships(user)
+                purger.delete_group_creators(user)
+                purger.delete_user(user)
+            except LimitReached:
+                break
+            else:
+                completed_jobs.append(job)
+
+        self.job_queue.delete(completed_jobs)
+
+
+class UserPurger:
+    """Helper methods for purging data belonging to a given user."""
+
+    def __init__(self, db, job_queue, worker):
+        self.db = db
+        self.job_queue = job_queue
+        self.worker = worker
+
+    def delete_authtickets(self, user):
+        """Delete all AuthTicket's belonging to `user`."""
+        self.worker.delete(
+            AuthTicket, select(AuthTicket.id).where(AuthTicket.user == user)
+        )
+
+    def delete_tokens(self, user):
+        """Delete all tokens belonging to `user`."""
+        self.worker.delete(Token, select(Token.id).where(Token.user == user))
+
+    def delete_flags(self, user):
+        """Delete all flags created by `user`."""
+        self.worker.delete(Flag, select(Flag.id).where(Flag.user_id == user.id))
+
+    def delete_featurecohort_memberships(self, user):
+        """Remove `user` from all feature cohorts."""
+        self.worker.delete(
+            FeatureCohortUser,
+            select(FeatureCohortUser.id).where(FeatureCohortUser.user_id == user.id),
+        )
```

Comment:
I think all these methods may belong in model-specific services (`AuthTicketService` etc) where they can potentially be re-used by other code. But that's a refactoring that I don't want to bite off right now.